File: app/src/main/java/com/zwstudio/logicgamesandroid/bridges/domain/BridgesGame.java
package com.zwstudio.logicgamesandroid.bridges.domain;

import com.zwstudio.logicgamesandroid.logicgames.domain.CellsGame;
import com.zwstudio.logicgamesandroid.logicgames.domain.GameInterface;
import com.zwstudio.logicgamesandroid.logicgames.domain.Position;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by zwvista on 2016/09/29.
 */

public class BridgesGame extends CellsGame<BridgesGame, BridgesGameMove, BridgesGameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(0, -1),
    };

    public Map<Position, BridgesIslandInfo> islandsInfo = new HashMap<>();
    public boolean isIsland(Position p) {return islandsInfo.containsKey(p);}

    public BridgesGame(List<String> layout, GameInterface<BridgesGame, BridgesGameMove, BridgesGameState> gi) {
        super(gi);
        size = new Position(layout.size(), layout.get(0).length());
        BridgesGameState state = new BridgesGameState(this);
        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                char ch = str.charAt(c);
                if (ch >= '0' && ch <= '9') {
                    BridgesIslandInfo info = new BridgesIslandInfo();
                    info.bridges = ch - '0';
                    islandsInfo.put(p, info);
                    state.set(r, c, new BridgesIslandObject());
                } else
                    state.set(r, c, new BridgesEmptyObject());
            }
        }
        for (Map.Entry<Position, BridgesIslandInfo> entry : islandsInfo.entrySet()) {
            Position p = entry.getKey();
            BridgesIslandInfo info = entry.getValue();
            for (int i = 0; i < 4; i++) {
                Position os = offset[i];
                for (Position p2 = p.add(os); isValid(p2); p2.addBy(os))
                    if (isIsland(p2)) {
                        info.neighbors[i] = p2;
                        break;
                    }
            }
        }
        states.add(state);
        levelInitilized(state);
    }

    public boolean switchBridges(Position pFrom, Position pTo) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        BridgesGameState state = cloner.deepClone(state());
        if (pTo.compareTo(pFrom) < 0) {
            Position t = pFrom;
            pFrom = pTo;
            pTo = t;
        }
        BridgesGameMove move = new BridgesGameMove();
        move.pFrom = pFrom; move.pTo = pTo;
        if (!state.switchBridges(move)) return false;
        states.add(state);
        stateIndex++;
        moves.add(move);
        moveAdded(move);
        levelUpdated(states.get(stateIndex - 1), state);
        return true;
   }

    public BridgesObject getObject(Position p) {
        return state().get(p);
    }

    public BridgesObject getObject(int row, int col) {
        return state().get(row, col);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/bridges/BridgesGame.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.bridges

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class BridgesGame(layout: List<String>, gi: GameInterface<BridgesGame, BridgesGameMove, BridgesGameState>, gdi: GameDocumentInterface) : CellsGame<BridgesGame, BridgesGameMove, BridgesGameState>(gi, gdi) {
    companion object {
        var offset = arrayOf(
            Position(-1, 0),
            Position(0, 1),
            Position(1, 0),
            Position(0, -1)
        )
    }

    var islandsInfo = mutableMapOf<Position, BridgesIslandInfo>()
    fun isIsland(p: Position) = islandsInfo.containsKey(p)

    init {
        size = Position(layout.size, layout[0].length)
        for (r in 0 until rows) {
            val str = layout[r]
            for (c in 0 until cols) {
                val p = Position(r, c)
                val ch = str[c]
                if (ch in '0'..'9') {
                    val info = BridgesIslandInfo()
                    info.bridges = ch - '0'
                    islandsInfo[p] = info
                }
            }
        }
        for ((p, info) in islandsInfo) {
            for (i in 0 until 4) {
                val os = offset[i]
                var p2 = p + os
                while (isValid(p2)) {
                    if (isIsland(p2)) {
                        info.neighbors[i] = p2
                        break
                    }
                    p2 += os
                }
            }
        }
        val state = BridgesGameState(this)
        levelInitilized(state)
    }

    fun switchBridges(move: BridgesGameMove) =
        changeObject(move) { state, move ->
            if (move.pTo < move.pFrom) {
                val t = move.pFrom
                move.pFrom = move.pTo
                move.pTo = t
            }
            state.switchBridges(move)
        }

    override fun setObject(move: BridgesGameMove) = switchBridges(move)

    fun getObject(p: Position) = currentState[p]
    fun getObject(row: Int, col: Int) = currentState[row, col]
}